### PR TITLE
[IMP] Invoicing improvements

### PR DESCRIPTION
**Pull request by Megaaaaaa**
*: account, l10n_be, web

---

This PR contains multiple commit which all serve a common purpose which is making invoicing clearer and more friendly to the user.

The commits that compose this PR are three. A short explanation for each of them will be provided here but more information can be found in their respective commit messages.

1. Enhancing the UI by adding, moving and removing some fields and displayed text. This is mostly cosmetic.
2. Enhancing the UX by modifying the behavior of already selected records in many-to-many tags fields (like taxes for example)
3. Modifying the Many2Many widget for taxes only to rename the search more option and display the tax_scope in the autocomplete

---

Enterprise PR: https://github.com/odoo/enterprise/pull/61327
task-3642419

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account/static/src/components/tax_autocomplete/tax_autocomplete.scss [hunk 1]
```diff
@@ -0,0 +1,4 @@
+.tax_autocomplete_grid {
+    display: grid;
+    grid-template-columns: minmax(auto, 1fr) minmax(auto, 1fr);
+}
```
Comment (FlorianGilbert) [inline addons/account/static/src/components/tax_autocomplete/tax_autocomplete.scss:4]: Empty line at the end of the file

File: addons/web/static/src/views/fields/relational_utils.js [hunk 2]
```diff
@@ -317,8 +324,9 @@ export class Many2XAutocomplete extends Component {
         if (this.lastProm) {
             this.lastProm.abort(false);
         }
-        this.lastProm = this.search(request);
-        const records = await this.lastProm;
+        const searchPromise = this.abortableSearch(request);
+        this.lastProm = searchPromise;
+        const records = await searchPromise.promise;
```
Comment (FlorianGilbert) [inline addons/web/static/src/views/fields/relational_utils.js:329]: Why don't we do that:

```suggestion
        this.lastProm = this.abortableSearch(request);
        const records = await this.lastProm.promise;
```

**Review by FlorianGilbert — CHANGES_REQUESTED** [review]
Hello

There is a typo in your first commit:

> The width of certain columns which were **preivously** using default values have
`preivously` should be `previously`

Another thing in your second commit message:

> A **traduction** test on the tax description has been adapted.
`translation` instead of `traduction`
Btw, concerning this commit, I will squash it with the first one because it's fixing tests that are broken by the first commit so as we don't want to merge a broken commit, let's squash it together.

In the fourth commit, we explain the label change but not the change on the search, could you please explain in few word why we're doing that, thank you

Another typo in the fifth commit:

> The way the tax scope is shown in the tax many2many dropdown **meny** is not clear
and can be improved.

`menu` instead of `meny` ?

Same for the last commit concerning the squash, as you're fixing tests there, please squash this commit with the broken commit, thank you

File: addons/web/static/src/views/fields/many2many_tags/many2many_tags_field.js [hunk 3]
```diff
@@ -158,19 +166,19 @@ export class Many2ManyTagsField extends Component {
...
-        const tagRecord = this.props.record.data[this.props.name].records.find(
-            (record) => record.id === id
-        );
-        await this.props.record.data[this.props.name].forget(tagRecord);
+        let tagRecord;
+        while (
+            (tagRecord = this.props.record.data[this.props.name].records.find(
+                (record) => record.id === id
+            ))
+        ) {
```
[the remark this answers is not in the document]
Comment (Megaaaaaa) [reply addons/web/static/src/views/fields/many2many_tags/many2many_tags_field.js:174]: This was more a safety net than something actually required. It is true it makes only very little sense for something like that so I will simply remove it.

File: addons/web/static/src/views/fields/many2many_tags/many2many_tags_field.js [hunk 4]
```diff
@@ -143,9 +146,14 @@ export class Many2ManyTagsField extends Component {
     }
 
     get tags() {
-        return this.props.record.data[this.props.name].records.map((record) =>
-            this.getTagProps(record)
-        );
+        // Use Set to remove client-side duplicates of selected tags, ensuring each tag is displayed only once.
+        return [
+            ...new Set(
+                this.props.record.data[this.props.name].records.map((record) =>
```
Comment (aab-odoo) [inline addons/web/static/src/views/fields/many2many_tags/many2many_tags_field.js:152]: there should be no duplicate in this.props.record.data[this.props.name].records (the model is the source of truth, it should be correct)

File: addons/web/static/src/views/fields/many2many_tags/many2many_tags_field.js [hunk 5]
```diff
@@ -158,19 +166,18 @@ export class Many2ManyTagsField extends Component {
     }
 
     async deleteTag(id) {
-        const tagRecord = this.props.record.data[this.props.name].records.find(
+        let tagRecord = this.props.record.data[this.props.name].records.find(
             (record) => record.id === id
         );
-        await this.props.record.data[this.props.name].forget(tagRecord);
+
+        if (tagRecord) {
```
Comment (aab-odoo) [inline addons/web/static/src/views/fields/many2many_tags/many2many_tags_field.js:173]: I'm not sure to understand why/how you could remove a tag that isn't in the relation

File: addons/web/static/src/views/fields/many2many_tags/many2many_tags_field.js [hunk 6]
```diff
@@ -99,6 +99,9 @@ export class Many2ManyTagsField extends Component {
             if (!recordlist) {
                 return;
             }
+            recordlist = recordlist.filter((element) => {
```
Comment (aab-odoo) [inline addons/web/static/src/views/fields/many2many_tags/many2many_tags_field.js:102]: besides the removal of the domain (diff in getDomain), this is the only other diff that should be there, I think (i.e. you don't filter out records that are already in the relation, and when the user clicks on them, you make sur we don't add them a second time to the relation). The remaining diff shouldn't exist (or I'm missing something)

Comment (aab-odoo) [reply addons/web/static/src/views/fields/many2many_tags/many2many_tags_field.js:102]: Also, the new behavior should be tested (I'm surprised no tests fail, actually)

File: addons/account/tests/test_tour.py [hunk 7]
```diff
@@ -80,3 +80,6 @@ def test_01_account_tax_groups_tour(self):
         product.supplier_taxes_id = new_tax
 
         self.start_tour("/web", 'account_tax_group', login="admin")
+
+    def test_01_account_tax_widget_tour(self):
+        self.start_tour("/web", 'account_tax_widget_tour', login="admin")
```
Comment (aab-odoo) [inline addons/account/tests/test_tour.py:85]: I meant unit tests in web/.

**Comment by Megaaaaaa** [comment]
Hello @odoo/rd-security :wave: 

Could you review this PR please ?

A previously Char field is changed for an Html field and so the tests which used to look for a simple text string now have to look for a Markup text.

No input is ever used inside the new Markup() methods so i believe it makes them safe.

Thank you in advance :pray:

File: addons/web/static/tests/legacy/views/fields/many2many_tags_avatar_field_tests.js [hunk 8]
```diff
@@ -399,24 +399,25 @@ QUnit.module("Fields", (hooks) => {
         );
         // select first input
         await click(popover.querySelector(".o-autocomplete--dropdown-item"));
-        assert.strictEqual(popover.querySelectorAll(".o_tag").length, 3, "Should have 3 tags");
+        // first input is already selected so it should remain two tags
+        assert.strictEqual(popover.querySelectorAll(".o_tag").length, 2, "Should have 2 tags");
```
Comment (aab-odoo) [inline addons/web/static/tests/legacy/views/fields/many2many_tags_avatar_field_tests.js:403]: You change the semantics of this test, I would rather click on the third (?) element (the one that was first before your change), to keep the exact same semantics.

**Review by aab-odoo — APPROVED** [review]
Besides the remaining comment, I approve the change on many2many tags field

**Review by FlorianGilbert — APPROVED** [review]
@robodoo r+

**Comment by FlorianGilbert** [comment]
@robodoo rebase-ff

**Comment by seb-odoo** [comment]
Hello, what was the reasoning to show already selected tags on m2m autocomplete?

I don't know why it would be useful in accounting, but for me this is a big annoyance, especially in project.

Could it be made as a field specific option instead, only for those that need this strange behavior?

If a tag is already set, I don't want to see it again or set it again (as it has no effect, as stated in the commit), so it doesn't need to suggest it. It's confusing and making search results worse, as we only display few of them, the few available spots are wasted. This forces users to use more specific search term to find relevant results, whereas very few chars could be used when already selected tags were filtered out.

And to avoid confusion and clicks that do nothing, if we _really_ must keep them, it should at least display something next to the value to tell it was already selected, and even possibly "disable" the line.

**Comment by chklop** [comment]
@seb-odoo This discussion has already sailed, please see the discussion in https://www.odoo.com/odoo/project.task/3919307

**Pull request closed without merging**